### PR TITLE
[CoreML EP] Add support of BatchNormalization/Reshape/Global[Average/Max]Pool

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: [CoreML EP] Add support of BatchNormalization/Reshape/Global[Average/Max]Pool

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/providers/coreml/builders/impl/batch_norm_op_builder.cc [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_gamma(), scale_tensor);   // scale
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_beta(), bias_tensor);     // B
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_mean(), mean_tensor);     // mean
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_variance(), var_tensor);  // var
+
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(layer.release());
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/impl/batch_norm_op_builder.cc:75]: >AddLayer [](start = 16, length = 8)

would it be better if ModelBuilder::AddLayer() takes a unique_ptr? then it conveys the transfer of ownership too

File: onnxruntime/core/providers/coreml/builders/impl/batch_norm_op_builder.cc [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    return false;
+
+  const auto input_size = input_shape.size();
+  // TODO, support 1d batch normalization (input is 3d)
+  if (input_size != 4) {
+    LOGS(logger, VERBOSE) << "BN only support 4d shape for now, input is "
+                          << input_size << "d shape";
+    return false;
+  }
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/impl/batch_norm_op_builder.cc:101]: Can we infer the batch size to be 1 and treat it as 4d if input_size == 3?

File: onnxruntime/core/providers/coreml/builders/impl/pool_op_builder.cc [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  // We only support global pool now
+  coreml_pool->set_globalpooling(true);
+  coreml_pool->mutable_valid();
+
+  if (op_type == "GlobalAveragePool") {
+    coreml_pool->set_type(COREML_SPEC::PoolingLayerParams_PoolingType_AVERAGE);
+  } else if (op_type == "GlobalMaxPool") {
+    coreml_pool->set_type(COREML_SPEC::PoolingLayerParams_PoolingType_MAX);
+  }
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/impl/pool_op_builder.cc:45]: Should there be an 'else' branch here to catch unexpected usage?

File: onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const logging::Logger& logger) const override;
+
+  // Reshape opset 4- uses attributes for new shape which we do not support for now
+  int GetMinSupportedOpSet(const Node& /* node */) const override { return 5; }
+};
```
Comment (skottmckay) [inline onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc:33]: FWIW opset 7 was the first opset ORT ever supported so anything prior to that is irrelevant

File: onnxruntime/core/providers/coreml/builders/impl/batch_norm_op_builder.cc (hunk 2 above)
Comment (gwang-msft) [reply onnxruntime/core/providers/coreml/builders/impl/batch_norm_op_builder.cc:101]: This will be handled in a future PR, will need to add 1 to the end of the shape and make this a 2d batch norm, such as {N,C,H} -> {N,C,H,1}
Will need to add a ExpandDimsLayer (unsqueeze) before bn and a SqueezeLayer after
This may also be required by other ops since CoreML has weird shape for some layers

---
In reply to: [573438207](https://github.com/microsoft/onnxruntime/pull/6625#discussion_r573438207) [](ancestors = 573438207)

File: onnxruntime/core/providers/coreml/builders/impl/pool_op_builder.cc (hunk 3 above)
Comment (gwang-msft) [reply onnxruntime/core/providers/coreml/builders/impl/pool_op_builder.cc:45]: Add it to the IsOpSupportedImpl() so this will never happen here

File: onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc (hunk 4 above)
Comment (gwang-msft) [reply onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc:33]: There are some op is the same since opset 1, for simplicity, keep it this way

**Review by skottmckay — APPROVED** [review]
